### PR TITLE
Add outlier in AWQ test cases

### PR DESCRIPTION
**Pull request by namgyu-youn**
**Summary**:
Similar to #3101 , this test inject outliers to reflect input distribution. Toy model is updated to use direct dtype, device callsite

**Test plan**:
pytest -sv test/prototype/test_awq.py

### CODE REVIEW

**Comment by namgyu-youn** [comment]
@Xia-Weiwen could you please review this PR?

**Review by Xia-Weiwen — APPROVED** [review]
LGTM

**Review by jerryzh168 — APPROVED** [review]

**Comment by jerryzh168** [comment]
@namgyu-youn if you have time, it will be very helpful if you can help us debug why the unittest doesn't work for hqq: here we are using the default, not hqq currently: https://github.com/pytorch/ao/blob/2fe0ca0899c730c528efdbec8886feaa38879f39/test/prototype/test_awq.py#L63

just need to update / or add a test for `Int4WeightOnlyConfig(group_size=128, int4_packing_format="tile_packed_to_4d", int4_choose_qparams_algorithm="hqq")`

**Comment by namgyu-youn** [comment]
> @namgyu-youn if you have time, it will be very helpful if you can help us debug why the unittest doesn't work for hqq: here we are using the default, not hqq currently:
> 
> https://github.com/pytorch/ao/blob/2fe0ca0899c730c528efdbec8886feaa38879f39/test/prototype/test_awq.py#L63
> 
> just need to update / or add a test for `Int4WeightOnlyConfig(group_size=128, int4_packing_format="tile_packed_to_4d", int4_choose_qparams_algorithm="hqq")`

Just want to make it clear, could you explain how mixed (AWQ + HQQ) PTQ works at a high level? They look like opposite sides: calibration-based vs. calibration-free. Is it work like:
1. Scale weight & activation (AWQ with calibration) 
2. HQQ simulation: Compute error
(loop 1->2)
3. Quantization with best scale using real HQQ
 
?

btw, SINQ (https://huggingface.co/papers/2509.22944) looks very interesting. It is a calibration-free PTQ like HQQ, and uses **2-axis (row/column) scale factor** to efficient outlier observation

**Comment by jerryzh168** [comment]
> > @namgyu-youn if you have time, it will be very helpful if you can help us debug why the unittest doesn't work for hqq: here we are using the default, not hqq currently:
> > https://github.com/pytorch/ao/blob/2fe0ca0899c730c528efdbec8886feaa38879f39/test/prototype/test_awq.py#L63
> > 
> > just need to update / or add a test for `Int4WeightOnlyConfig(group_size=128, int4_packing_format="tile_packed_to_4d", int4_choose_qparams_algorithm="hqq")`
> 
> Just want to make it clear, could you explain how mixed (AWQ + HQQ) PTQ works at a high level? They look like opposite sides: calibration-based vs. calibration-free. Is it work like:
> 
> 1. Scale weight & activation (AWQ with calibration)
> 2. HQQ simulation: Compute error
>    (loop 1->2)
> 3. Quantization with best scale using real HQQ
> 
> ?
> 
> btw, SINQ ([huggingface.co/papers/2509.22944](https://huggingface.co/papers/2509.22944), [arxiv.org/abs/2509.22944](https://arxiv.org/abs/2509.22944)) looks very interesting I feel. SINQ is a calibration-free PTQ like HQQ. and uses a **2-axis (row/column) scale factor** to efficient outlier observation

oh so hqq just replaces the original naive choose_affine_qparams with its own algorithm to choose_qparams, see https://github.com/pytorch/ao/blob/28612d00eaa711ba91e23d0aa9aec3ca45ba37b2/torchao/quantization/quantize_/workflows/int4/int4_tile_packed_to_4d_tensor.py#L148

and awq is applying an additional pre_scaling_factor on top of the scale/zero_point for quantization

so the process is:
```
# AWQ loop to search for pre_scale factor for activation and weight
for act_pre_scale in scale_factor_options:
    weight = original_weight * act_pre_scale
    qw = quantize_with_hqq(weight)
    activation = original_activation / act_pre_scale
    awq_result = F.linear(activation, qw)
    ref_result = F.linear(original_activation, original_weight)
    loss = loss_fn(ref_result, awq_result)
    ... # find scale that give the best loss
```

I think it's a problem because in principle awq should not regress the metric that we are measuring: https://github.com/pytorch/ao/blob/28612d00eaa711ba91e23d0aa9aec3ca45ba37b2/test/prototype/test_awq.py#L130-L131

**Comment by jerryzh168** [comment]
SINQ seems to be pretty new and looks like it has good improvement in lower bits, contributions are welcome as well

**Comment by namgyu-youn** [comment]
> SINQ seems to be pretty new and looks like it has good improvement in lower bits, contributions are welcome as well

Yeah SINQ (https://github.com/huawei-csl/SINQ) is pretty new, but 2-axis scale factor (fig1) looked pretty cool in my first look.

image

If we can reuse weight-only api (i.e., only implement `choose_qparams_sinq`), 2-axis scale factor seems to be a choice in my thought. But not now yet, because I am learning Marlin kernel implementation.

**Comment by namgyu-youn** [comment]
> @namgyu-youn if you have time, it will be very helpful if you can help us debug why the unittest doesn't work for hqq: here we are using the default, not hqq currently:
> 
> https://github.com/pytorch/ao/blob/2fe0ca0899c730c528efdbec8886feaa38879f39/test/prototype/test_awq.py#L63
> 
> just need to update / or add a test for `Int4WeightOnlyConfig(group_size=128, int4_packing_format="tile_packed_to_4d", int4_choose_qparams_algorithm="hqq")`

Please check https://github.com/pytorch/ao/pull/3148 to fix this issue, thanks.
